Argument list too long from a coq_makefile-generated Makefile

I'll soon package Coq 8.17 for Debian (we're in a freeze) ; for that I started checking all Coq-related packages for correct support.

The case of unimath 20230420 is interesting: it does compile, but at the install stage:
```
ulimit -v unlimited ; make -f build/CoqMakefile.make install
make[3]: Entering directory '/<<PKGBUILDDIR>>'
make[3]: /bin/sh: Argument list too long
make[3]: *** [build/CoqMakefile.make:598: install] Error 127
make[3]: Leaving directory '/<<PKGBUILDDIR>>'
make[2]: *** [Makefile:97: install] Error 2
make[2]: Leaving directory '/<<PKGBUILDDIR>>'
make[1]: *** [debian/rules:18: override_dh_auto_install] Error 2
make[1]: Leaving directory '/<<PKGBUILDDIR>>'
make: *** [debian/rules:6: binary] Error 2
```

The install target in build/CoqMakefile.make is:
```
install: META
        $(HIDE)code=0; for f in $(FILESTOINSTALL); do\
         if ! [ -f "$$f" ]; then >&2 echo $$f does not exist; code=1; fi \
        done; exit $$code
        $(HIDE)for f in $(FILESTOINSTALL); do\
         df="`$(COQMKFILE) -destination-of "$$f" $(COQLIBS)`";\
         if [ "$$?" != "0" -o -z "$$df" ]; then\
           echo SKIP "$$f" since it has no logical path;\
         else\
           install -d "$(COQLIBINSTALL)/$$df" &&\
           install -m 0644 "$$f" "$(COQLIBINSTALL)/$$df" &&\
           echo INSTALL "$$f" "$(COQLIBINSTALL)/$$df";\
         fi;\
        done
        $(call findlib_remove)
        $(call findlib_install, META $(FINDLIBFILESTOINSTALL))
        $(HIDE)$(MAKE) install-extra -f "$(SELF)"
```

I poked around and found [this discussion](https://stackoverflow.com/questions/64201512/argument-list-too-long-when-linking-with-gnu-make) mentioning a potential fix using [this trick](https://www.gnu.org/software/make/manual/html_node/File-Function.html#File-Function).

I'm not 100% sure my diagnosis is good, but there something odd and I think it's the right place to report.

--- a/doc/changelog/09-cli-tools/17697-filestoinstall.rst
+++ b/doc/changelog/09-cli-tools/17697-filestoinstall.rst
@@ -1,0 +1,6 @@
+- **Fixed:**
+  `coq_makefile` avoids generating a command containing all files to install in a make rule,
+  which could surpass the maximum single argument size in some developments
+  (`#17697 <https://github.com/coq/coq/pull/17697>`_,
+  fixes `#17721 <https://github.com/coq/coq/issues/17721>`_,
+  by Gaëtan Gilbert).

--- a/tools/CoqMakefile.in
+++ b/tools/CoqMakefile.in
@@ -592,13 +592,27 @@ beautify: $(BEAUTYFILES)
 # There rules can be extended in @LOCAL_FILE@
 # Extensions can't assume when they run.
 
+# We use $(file) to avoid generating a very long command string to pass to the shell
+# (cf https://coq.zulipchat.com/#narrow/stream/250632-Coq-Platform-devs-.26-users/topic/Strange.20command.20length.20limit.20on.20Linux)
+# However Apple ships old make which doesn't have $(file) so we need a fallback
+$(file >.hasfile,1)
+HASFILE:=$(shell if [ -e .hasfile ]; then echo 1; rm .hasfile; fi)
+
+.filestoinstall:
+	@:$(if $(HASFILE),$(file >$@,$(FILESTOINSTALL)),\
+	  $(shell rm -f $@) \
+	  $(foreach x,$(FILESTOINSTALL),$(shell printf '%s\n' "$x" >> $@)))
+
+
+.PHONY: .filestoinstall
+
 # findlib needs the package to not be installed, so we remove it before
 # installing it (see the call to findlib_remove)
-install: META
-	$(HIDE)code=0; for f in $(FILESTOINSTALL); do\
+install: META .filestoinstall
+	$(HIDE)code=0; for f in $$(cat .filestoinstall); do\
 	 if ! [ -f "$$f" ]; then >&2 echo $$f does not exist; code=1; fi \
 	done; exit $$code
-	$(HIDE)for f in $(FILESTOINSTALL); do\
+	$(HIDE)for f in $$(cat .filestoinstall); do\
 	 df="`$(COQMKFILE) -destination-of "$$f" $(COQLIBS)`";\
 	 if [ "$$?" != "0" -o -z "$$df" ]; then\
 	   echo SKIP "$$f" since it has no logical path;\
@@ -640,16 +654,16 @@ install-doc:: html mlihtml
 	done
 .PHONY: install-doc
 
-uninstall::
+uninstall:: .filestoinstall
 	@# Extension point
 	$(call findlib_remove)
-	$(HIDE)for f in $(FILESTOINSTALL); do \
+	$(HIDE)for f in $$(cat .filestoinstall); do \
 	 df="`$(COQMKFILE) -destination-of "$$f" $(COQLIBS)`" &&\
 	 instf="$(COQLIBINSTALL)/$$df/`basename $$f`" &&\
 	 rm -f "$$instf" &&\
 	 echo RM "$$instf" ;\
 	done
-	$(HIDE)for f in $(FILESTOINSTALL); do \
+	$(HIDE)for f in $$(cat .filestoinstall); do \
 	 df="`$(COQMKFILE) -destination-of "$$f" $(COQLIBS)`" &&\
 	 echo RMDIR "$(COQLIBINSTALL)/$$df/" &&\
 	 (rmdir "$(COQLIBINSTALL)/$$df/" 2>/dev/null || true); \

